chainindex/read.go

```diff
@@ -0,0 +1,170 @@
+package chainindex
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/types/ethtypes"
+)
+
+var (
+	headIndexedWaitTimeout = 5 * time.Second
+)
+
+func (si *SqliteIndexer) GetMaxNonRevertedTipset(ctx context.Context) (*types.TipSet, error) {
+	si.closeLk.RLock()
+	if si.closed {
+		return nil, ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	var tipsetKeyCidBytes []byte
+	err := si.getMaxNonRevertedTipsetStmt.QueryRowContext(ctx).Scan(&tipsetKeyCidBytes)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, nil
+		}
+		return nil, xerrors.Errorf("failed to get max non reverted tipset: %w", err)
+	}
+
+	tipsetKeyCid, err := cid.Cast(tipsetKeyCidBytes)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to cast tipset key cid: %w", err)
+	}
+
+	// Can this error out for reverted tipsets ?
+	return si.cs.GetTipSetByCid(ctx, tipsetKeyCid)
+}
+
+func (si *SqliteIndexer) GetCidFromHash(ctx context.Context, txHash ethtypes.EthHash) (cid.Cid, error) {
+	si.closeLk.RLock()
+	if si.closed {
+		return cid.Undef, ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	var msgCidBytes []byte
+
+	err := si.queryMsgCidFromEthHash(ctx, txHash, &msgCidBytes)
+	if err == sql.ErrNoRows {
+		err = si.waitTillHeadIndexedAndApply(ctx, func() error {
+			return si.queryMsgCidFromEthHash(ctx, txHash, &msgCidBytes)
+		})
+	}
+
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return cid.Undef, ErrNotFound
+		}
+		return cid.Undef, xerrors.Errorf("failed to get message CID from eth hash: %w", err)
+	}
+
+	msgCid, err := cid.Cast(msgCidBytes)
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to cast message CID: %w", err)
+	}
+
+	fmt.Println("RETURNING CORRECT MSG CID")
+
+	return msgCid, nil
+}
+
+func (si *SqliteIndexer) queryMsgCidFromEthHash(ctx context.Context, txHash ethtypes.EthHash, msgCidBytes *[]byte) error {
+	return si.getMsgCidFromEthHashStmt.QueryRowContext(ctx, txHash.String()).Scan(msgCidBytes)
+}
+
+func (si *SqliteIndexer) GetMsgInfo(ctx context.Context, messageCid cid.Cid) (*MsgInfo, error) {
+	si.closeLk.RLock()
+	if si.closed {
+		return nil, ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	var tipsetKeyCidBytes []byte
+	var height int64
+
+	err := si.queryMsgInfo(ctx, messageCid, &tipsetKeyCidBytes, &height)
+	if err == sql.ErrNoRows {
+		err = si.waitTillHeadIndexedAndApply(ctx, func() error {
+			return si.queryMsgInfo(ctx, messageCid, &tipsetKeyCidBytes, &height)
+		})
+	}
+
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, ErrNotFound
+		}
+		return nil, xerrors.Errorf("failed to get message info: %w", err)
+	}
```

Comment:
I have a cognitive speed-bump at these lines, as in the above function and I wonder if you could abstract this into a function to handle both cases -- provide a getter and have the function run it, then if it fails, wait and run it again; is that do-able?

Reply:
then your getter wouldn't even need to be a method on `SqliteIndexer`, it could just be a simple closure

Reply:
Agreed. Refactored this to make it easy to reason about.

Reply:
nice